### PR TITLE
Bug fixes

### PR DESCRIPTION
**Pull request by apaszke**
* `torch.Size` methods convert return type from `tuple` back to `torch.Size` (#730)
* Fixed bug in `Engine::compute_dependencies` (#726)
* `detach()` now actually removes the creator (#685)
* Added `torch.__version__` (#742)

### CODE REVIEW

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -1,33 +1,72 @@
...
+      } else if (reachable && visit.fn->requires_grad && prev_fn->requires_grad) {
+        dependencies[prev_fn] += 1;
       }
-      if (seen.count(prev_fn.get()) == 0) {
-        seen.insert(prev_fn.get());
-        queue.push_back(prev_fn);
+      auto it = seen.find(prev_fn);
+      if (it == seen.end() || (it->second == false && reachable)) {
+        seen[prev_fn] = reachable;
+        queue.emplace(prev_fn, reachable);
```
Comment (colesbury) [inline torch/csrc/autograd/engine.cpp]: This looks like the same function can be put in the queue multiple times. Is that safe?

File: torch/autograd/variable.py [hunk 2]
```diff
@@ -212,8 +212,24 @@ def reinforce(self, reward):
...
+          in-place modifications on either of them will be seen, and may trigger
+          errors in correctness checks.
+        """
+        result = NoGrad()(self)  # this is needed, because it merges version counters
+        result._creator = None
+        return result
+
+    def detach_(self):
+        """Detaches the Variable from the graph that created it, making it a leaf."""
+        self._creator = None
```
Comment (apaszke) [inline torch/autograd/variable.py]: I don't have a strong preference, but if we're going to separate being a leaf and not having a creator, then it might make sense to change it too.

File: setup.py [hunk 3]
```diff
@@ -362,10 +369,18 @@ def make_relative_rpath(path):
                        )
     extensions.append(THCUNN)
 
-version = "0.1"
+version = '0.1'
 if os.getenv('PYTORCH_BUILD_VERSION'):
```
Comment (apaszke) [inline setup.py]: oooook

File: torch/csrc/Size.cpp [hunk 4]
```diff
@@ -54,6 +54,43 @@ static PyObject * THPSize_repr(THPSize *self)
...
+static auto sq_concat = PyTuple_Type.tp_as_sequence->sq_concat;
+static auto sq_repeat = PyTuple_Type.tp_as_sequence->sq_repeat;
+static auto sq_item = PyTuple_Type.tp_as_sequence->sq_item;
+static auto mp_subscript = PyTuple_Type.tp_as_mapping->mp_subscript;
+
+static PySequenceMethods THPSize_as_sequence = {
+  PyTuple_Type.tp_as_sequence->sq_length,
+  wrap_tuple_fn<decltype(&sq_concat), &sq_concat>,
+  wrap_tuple_fn<decltype(&sq_repeat), &sq_repeat>,
+  wrap_tuple_fn<decltype(&sq_item), &sq_item>,
```
Comment (apaszke) [inline torch/csrc/Size.cpp]: Right. I'll remove that

File: torch/__init__.py [hunk 5]
```diff
@@ -21,6 +21,11 @@
     'ShortTensor', 'CharTensor', 'ByteTensor',
 ]
 
+try:
+    from .version import __version__
+except ImportError:
```
Comment (apaszke) [inline torch/__init__.py]: Hmm, on second thought it doesn't make sense to wrap it. I'll remove it.

**Review by colesbury — APPROVED** [review]
LGTM

**Comment by lukeyeager** [comment]
Did you mean to tag 1bdc28161a68d24ab21d05b195935ef81c67826e as v0.1.9? There's a release branch but no tag.

**Comment by soumith** [comment]
@lukeyeager done. sorry for the delay: https://github.com/pytorch/pytorch/releases/tag/v0.1.9

**Comment by lukeyeager** [comment]
No problem at all! I'm loving the fact that there are versions in PyTorch :grinning:
